src/rpc/blockchain.cpp

```diff
@@ -2585,6 +2587,239 @@ static RPCHelpMan scanblocks()
+            }
+            if (blockindexes_sorted.insert(pindex).second) {
+                if (!heights_seen.insert(pindex->nHeight).second) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Duplicate heights found; blockhashes must be along a single chain");
+                }
+            }
+        }
+    }
+
+    std::set<CScript> scripts_to_watch;
+
+    // Determine scripts to watch.
+    for (const UniValue& scanobject : request.params[1].get_array().getValues()) {
+        FlatSigningProvider provider;
+        std::vector<CScript> scripts = EvalDescriptorStringOrObject(scanobject, provider);
+
+        for (const CScript& script : scripts) {
+            scripts_to_watch.insert(script);
+        }
+    }
+
+    const auto AddSpend = [&](
+            const CScript& spk,
+            const CAmount val,
+            const CTransactionRef& tx,
+            int vin,
+            const CTxIn& txin,
+            const CBlockIndex* index
+            ) {
+        UniValue event(UniValue::VOBJ);
+        UniValue spkUv(UniValue::VOBJ);
+        ScriptToUniv(spk, /*out=*/spkUv, /*include_hex=*/true, /*include_address=*/true);
+
+        event.pushKV("type", "spend");
+        event.pushKV("amount", ValueFromAmount(val));
+        event.pushKV("blockhash", index ? index->GetBlockHash().ToString() : "");
+        event.pushKV("height", index ? index->nHeight : -1);
+        event.pushKV("spend_txid", tx->GetHash().ToString());
+        event.pushKV("spend_vin", vin);
+        event.pushKV("prevout_txid", txin.prevout.hash.ToString());
+        event.pushKV("prevout_vout", txin.prevout.n);
+        event.pushKV("prevout_spk", spkUv);
+
+        return event;
+    };
+
+    const auto AddReceive = [&](const CTxOut& txout, const CBlockIndex* index, int vout, const CTransactionRef& tx) {
+        UniValue event(UniValue::VOBJ);
+        UniValue spkUv(UniValue::VOBJ);
+        ScriptToUniv(txout.scriptPubKey, /*out=*/spkUv, /*include_hex=*/true, /*include_address=*/true);
+
+        event.pushKV("type", "receive");
+        event.pushKV("amount", ValueFromAmount(txout.nValue));
+        event.pushKV("blockhash", index ? index->GetBlockHash().ToString() : "");
+        event.pushKV("height", index ? index->nHeight : -1);
+        event.pushKV("txid", tx->GetHash().ToString());
+        event.pushKV("vout", vout);
+        event.pushKV("output_spk", spkUv);
+
+        return event;
+    };
+
+    BlockManager* blockman;
+    Chainstate& active_chainstate = chainman.ActiveChainstate();
+    {
+        LOCK(::cs_main);
+        blockman = CHECK_NONFATAL(&active_chainstate.m_blockman);
+    }
+
+    for (const CBlockIndex* blockindex : blockindexes_sorted) {
+        const std::vector<uint8_t> block_data{GetRawBlockChecked(chainman.m_blockman, *blockindex)};
+        DataStream block_stream{block_data};
+        CBlock block{};
+        block_stream >> TX_WITH_WITNESS(block);
```

Comment:
Interesting edge case, agree it would be nice to show results for the blocks that can be found. Right now, passing in 1 available block and 1 pruned block in a pruned node results in an error.

```
➜  bitcoin git:(2024-08-getdescriptoractivity) ✗ bitcoinclimain getblockchaininfo
{
  "chain": "main",
  "blocks": 871458,
  "headers": 871458,
  "bestblockhash": "00000000000000000000f8ee686deb34ddfd6d888279a7f0b42f623f938745ea",
  "difficulty": 102289407543323.8,
  ...
  "pruned": true,
  "pruneheight": 870990,
  "automatic_pruning": true,
  "prune_target_size": 1073741824,
   ...
}
```

```
➜  bitcoin git:(2024-08-getdescriptoractivity) ✗ bitcoinclimain getdescriptoractivity \
'["00000000000000000001dca75f6da4a2ed1c24547e85e5a89fe75701726fb268","00000000000000000000e7b7e55b630178d7e001e0ed5aace4cb18f26ef5a2d9"]' \
'["addr(bc1qryhgpmfv03qjhhp2dj8nw8g4ewg08jzmgy3cyx)"]'
error code: -1
error message:
Block not available (pruned data)
```

Reply:
If this behaviour is incorporated, then a test testing with a present block and a pruned block would be nice as well.

And on similar lines, this test `test_invalid_blockhash` can also be updated wherein an invalid block hash and a valid one are passed but result is still returned for the valid one instead of the RPC throwing an error like it does atm. Below the first block is invalid, the second valid.

```
➜  bitcoin git:(2024-08-getdescriptoractivity) ✗ bitcoinclimain getdescriptoractivity \
'["00000000000000000001dca75f6da4a2ed1c24547e85e5a89fe75701726fb26a","00000000000000000000e7b7e55b630178d7e001e0ed5aace4cb18f26ef5a2d9"]' \
'["addr(bc1qryhgpmfv03qjhhp2dj8nw8g4ewg08jzmgy3cyx)"]'
error code: -5
error message:
Block not found
```